### ISSUE TITLE
Types narrowing at runtime

### ISSUE THREAD

**Issue by rarkins** [renovatebot/renovate#15746]
### Describe the proposed change(s).

I don't know if this is easily possible.

Let's say we have `RenovateConfig` type, which includes pretty much everything. Then we have narrower types like `ExtractConfig` which include much less. We can "narrow" in types from `RenovateConfig` to `ExtractConfig` if it's a subset, but the `config` object we pass is still just as big, so if we log it or we calculate a hash of its contents then it's still the same huge `RenovateConfig`.

Is there any solution existing which can let us narrow an object to a new type and pick only the applicable fields, without defining them in two locations?

**Comment by viceice** [comment]
at least there isn't a buildin typescript feature.

@zharinov maybe some of those validation libs you've tested before can do this now we mostly have strict null checks?

**Comment by zharinov** [comment]
If we have some function that operate on the small part of the big object, we can define it like `f(x: X): Y`, where `X` and `Y` is result of pick/omit helpers.

> @zharinov maybe some of those validation libs you've tested before can do this now we mostly have strict null checks?

We need it to be enabled entirely, though. The lib we agreed upon has more convenient `pick/omit` functionality compared to standard ones.

**Comment by astellingwerf** [comment]
Does https://stackoverflow.com/a/50895613/287820 tackle this?

**Comment by zharinov** [comment]
> Does https://stackoverflow.com/a/50895613/287820 tackle this?

I don't think we want to physically extract/move something, just improve the types

**Comment by rarkins** [comment]
It still requires essentially a "double" definition of types but at least it makes sure the two definitions stay in sync (I think)

**Comment by rarkins** [comment]
I want the actual object to be narrowed, so that it's more practical to log it as well as hash it's contents

**Comment by zharinov** [comment]
Another idea: redesign our "ultimate" data structure from scratch with our existing domain knowledge and while temporarily using both, gradually migrate to the new structure.

In my personal mental model, it's more like in-memory database than just big config, and we're kind of querying and writing to it.

**Comment by rarkins** [comment]
I'm not sure I like that idea, but also not sure I understand it..

**Comment by JamieMagee** [comment]
If I'm understanding it correctly, you want to drop all properties of the object that belong to the wider type?

It's not possible at runtime, because type information isn't available. The two ways I can think of doing it are:
- Using classes. Their properties are preserved at runtime
- Some sort of decorator to 'smuggle' the information through the ts compiler

**Comment by zharinov** [comment]
> I'm not sure I like that idea, but also not sure I understand it..

- Wrap the `RenovateConfig` behind the class
- Create methods that return narrow data
- Refactor `RenovateConfig` representation

**Comment by JamieMagee** [comment]
Yeah, I think this would be the only approach that would work and not duplicate types. Still not straightforward though.
